File: learning-kmp/bookshelf/iosApp/iosApp/DesignSystem/AppDialog.swift
import SwiftUI

struct AppDialog: View {
    let title: String
    let message: String
    let confirmLabel: String
    let onConfirm: () -> Void
    let onDismiss: () -> Void
    var confirmIsDestructive: Bool = false

    var body: some View {
        ZStack {
            Color.black.opacity(0.4)
                .ignoresSafeArea()
                .onTapGesture { onDismiss() }

            VStack(alignment: .leading, spacing: 12) {
                Text(title)
                    .font(.system(size: AppFontSize.titleMedium, weight: .bold))
                    .foregroundColor(AppColor.onSurface)

                Text(message)
                    .font(.system(size: AppFontSize.bodyMedium))
                    .foregroundColor(AppColor.onSurfaceVariant)

                Spacer().frame(height: 4)

                HStack(spacing: 12) {
                    Button(action: onDismiss) {
                        Text(AppStrings.cancel)
                            .font(.system(size: AppFontSize.bodyMedium, weight: .medium))
                            .foregroundColor(AppColor.onSurface)
                            .frame(maxWidth: .infinity)
                            .padding(.vertical, 14)
                            .background(AppColor.surfaceVariant)
                            .clipShape(Capsule())
                    }

                    Button(action: onConfirm) {
                        Text(confirmLabel)
                            .font(.system(size: AppFontSize.bodyMedium, weight: .medium))
                            .foregroundColor(
                                confirmIsDestructive ? AppColor.error : AppColor.primary
                            )
                            .frame(maxWidth: .infinity)
                            .padding(.vertical, 14)
                            .background(AppColor.surfaceVariant)
                            .clipShape(Capsule())
                    }
                }
            }
            .padding(24)
            .background(AppColor.surface)
            .clipShape(RoundedRectangle(cornerRadius: 16))
            .padding(.horizontal, 24)
        }
    }
}

File: learning-kmp/bookshelf/androidApp/src/main/kotlin/com/bookshelf/android/ui/AppDialog.kt
package com.bookshelf.android.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import com.bookshelf.design.AppStrings

@Composable
fun AppDialog(
    title: String,
    message: String,
    confirmLabel: String,
    onConfirm: () -> Unit,
    onDismiss: () -> Unit,
    confirmIsDestructive: Boolean = false
) {
    Dialog(onDismissRequest = onDismiss) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(16.dp))
                .background(MaterialTheme.colorScheme.surface)
                .padding(24.dp)
        ) {
            Text(
                text = title,
                style = MaterialTheme.typography.titleLarge,
                color = MaterialTheme.colorScheme.onSurface
            )
            Spacer(Modifier.height(12.dp))
            Text(
                text = message,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Spacer(Modifier.height(24.dp))
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                DialogButton(
                    label = AppStrings.CANCEL,
                    textColor = MaterialTheme.colorScheme.onSurface,
                    modifier = Modifier.weight(1f),
                    onClick = onDismiss
                )
                DialogButton(
                    label = confirmLabel,
                    textColor = if (confirmIsDestructive) MaterialTheme.colorScheme.error
                                else MaterialTheme.colorScheme.primary,
                    modifier = Modifier.weight(1f),
                    onClick = onConfirm
                )
            }
        }
    }
}

@Composable
private fun DialogButton(
    label: String,
    textColor: Color,
    modifier: Modifier = Modifier,
    onClick: () -> Unit
) {
    Box(
        modifier = modifier
            .clip(RoundedCornerShape(50.dp))
            .background(MaterialTheme.colorScheme.surfaceVariant)
            .clickable(onClick = onClick)
            .padding(vertical = 14.dp),
        contentAlignment = Alignment.Center
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.labelLarge,
            color = textColor
        )
    }
}
